core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -1007,6 +1007,33 @@ public Builder setBranchSnapshot(long snapshotId, String branch) {
       return this;
     }
 
+    public Builder setTag(String tag, SnapshotRef ref) {
+      SnapshotRef existingRef = refs.get(tag);
+      if (existingRef != null && existingRef.equals(ref)) {
+        return this;
+      }
+      long snapshotId = ref.snapshotId();
+      Snapshot snapshot = snapshotsById.get(snapshotId);
+      ValidationException.check(snapshot != null, "Cannot set %s to unknown snapshot: %s", tag, snapshotId);
+      ValidationException.check(ref.isTag(), "Cannot create tag: %s is a branch", tag);
+      refs.put(tag, ref);
+      return this;
+    }
+
+    public Builder setBranch(String branch, SnapshotRef ref) {
+      SnapshotRef existingRef = refs.get(branch);
+      if (existingRef != null && existingRef.equals(ref)) {
+        return this;
+      }
+      long snapshotId = ref.snapshotId();
+      Snapshot snapshot = snapshotsById.get(snapshotId);
+      ValidationException.check(snapshot != null, "Cannot set %s to unknown snapshot: %s", branch, snapshotId);
+      ValidationException.check(ref.isBranch(), "Cannot create branch: %s is a tag", branch);
+      refs.put(branch, ref);
+      return this;
```

Comment:
When writing tests, and debugging realized that I need to emit a change here. Will do that on next revision